File: src/legacy/unit_owned_type_4e7be0.go
package legacy

type unitOwnedTypeHooks4E7BE0[C, O comparable] struct {
	loadCache  func(C) uint32
	storeCache func(C, uint32)
	lookupType func(string) uint32
	loadFirst  func(O) O
	loadType   func(O) uint16
	loadNext   func(O) O
}

// unitHasOwnedType4E7BE0 preserves the shared owned-list contract in
// GAME.EXE 004E7BE0 and 004E7C30. The cache is loaded exactly once. A zero
// cache triggers lookup and store before the owner is dereferenced. Each
// object's zero-extended 16-bit type is compared with the complete 32-bit
// cached ID, and a match returns without loading that object's successor.
func unitHasOwnedType4E7BE0[C, O comparable](
	cache C,
	owner O,
	typeName string,
	hooks unitOwnedTypeHooks4E7BE0[C, O],
) uint32 {
	typeInd := hooks.loadCache(cache)
	if typeInd == 0 {
		typeInd = hooks.lookupType(typeName)
		hooks.storeCache(cache, typeInd)
	}

	obj := hooks.loadFirst(owner)
	var zero O
	for obj != zero {
		if uint32(hooks.loadType(obj)) == typeInd {
			return 1
		}
		obj = hooks.loadNext(obj)
	}
	return 0
}

func unitIsCrown4E7BE0[C, O comparable](cache C, owner O, hooks unitOwnedTypeHooks4E7BE0[C, O]) uint32 {
	return unitHasOwnedType4E7BE0(cache, owner, "Crown", hooks)
}

func unitIsGameBall4E7C30[C, O comparable](cache C, owner O, hooks unitOwnedTypeHooks4E7BE0[C, O]) uint32 {
	return unitHasOwnedType4E7BE0(cache, owner, "GameBall", hooks)
}

File: src/legacy/unit_owned_type_4e7be0.c
#include "GAME3_3.h"

enum {
	NOX_CROWN_TYPE_CACHE_4E7BE0 = 1567716,
	NOX_GAMEBALL_TYPE_CACHE_4E7C30 = 1567720,
};

static int nox_xxx_unitHasOwnedType_4E7BE0(const nox_object_t* owner, uint32_t* type_cache, char* type_name) {
	uint32_t type_ind = *type_cache;
	if (type_ind == 0) {
		type_ind = (uint32_t)nox_xxx_getNameId_4E3AA0(type_name);
		*type_cache = type_ind;
	}

	for (const nox_object_t* obj = owner->field_129; obj; obj = obj->field_128) {
		if ((uint32_t)obj->typ_ind == type_ind) {
			return 1;
		}
	}
	return 0;
}

//----- (004E7BE0) --------------------------------------------------------
int nox_xxx_unitIsCrown_4E7BE0(const nox_object_t* owner) {
	uint32_t* const type_cache = getMemU32Ptr(UINT32_C(0x5D4594), NOX_CROWN_TYPE_CACHE_4E7BE0);
	return nox_xxx_unitHasOwnedType_4E7BE0(owner, type_cache, "Crown");
}

//----- (004E7C30) --------------------------------------------------------
int nox_xxx_unitIsGameball_4E7C30(const nox_object_t* owner) {
	uint32_t* const type_cache = getMemU32Ptr(UINT32_C(0x5D4594), NOX_GAMEBALL_TYPE_CACHE_4E7C30);
	return nox_xxx_unitHasOwnedType_4E7BE0(owner, type_cache, "GameBall");
}
